ddsrouter_core/src/cpp/types/security/tls/TlsConfiguration.cpp

```diff
@@ -18,47 +18,167 @@
+#include <ddsrouter_utils/exception/InconsistencyException.hpp>
 
 namespace eprosima {
 namespace ddsrouter {
 namespace core {
 namespace types {
 namespace security {
 
+template <>
+void TlsConfiguration::check_valid_<TlsKind::client>() const
+{
+    if (certificate_authority_file_.empty())
+    {
+        // TODO check it is a correct file
+        throw utils::InitializationException(utils::Formatter() << "Invalid certificate_authority_file");
+    }
+}
+
+template <>
+void TlsConfiguration::check_valid_<TlsKind::server>() const
+{
+    // TODO check every file is correct
+    if ( private_key_file_.empty() or certificate_chain_file_.empty() or dh_params_file_.empty())
+    {
+        throw utils::InitializationException(utils::Formatter() << "At least one invalid file");
+    }
+}
+
+template <>
+void TlsConfiguration::check_valid_<TlsKind::both>() const
+{
+    this->check_valid_<TlsKind::client>();
+    this->check_valid_<TlsKind::server>();
+}
+
+// Inactive constructor
 TlsConfiguration::TlsConfiguration()
-    : kind_(TLS_INVALID)
+    : kind_(TlsKind::inactive)
 {
 }
 
+// Client constructor
 TlsConfiguration::TlsConfiguration(
-        TlsConfigurationKind kind)
-    : kind_(kind)
+        std::string certificate_authority_file)
+    : kind_(TlsKind::client)
+    , certificate_authority_file_(certificate_authority_file)
 {
+    this->check_valid_<TlsKind::client>();
 }
 
-bool TlsConfiguration::is_valid() const noexcept
+// Server constructor
+TlsConfiguration::TlsConfiguration(
+        std::string private_key_file_password,
+        std::string private_key_file,
+        std::string certificate_chain_file,
+        std::string dh_params_file)
+    : kind_(TlsKind::server)
+    , private_key_file_password_(private_key_file_password)
+    , private_key_file_(private_key_file)
+    , certificate_chain_file_(certificate_chain_file)
+    , dh_params_file_(dh_params_file)
 {
-    return true;
+    this->check_valid_<TlsKind::server>();
+}
+
+// Server & client constructor
+TlsConfiguration::TlsConfiguration(
+        std::string certificate_authority_file,
+        std::string private_key_file_password,
+        std::string private_key_file,
+        std::string certificate_chain_file,
+        std::string dh_params_file)
+    : kind_(TlsKind::both)
+    , certificate_authority_file_(certificate_authority_file)
+    , private_key_file_password_(private_key_file_password)
+    , private_key_file_(private_key_file)
+    , certificate_chain_file_(certificate_chain_file)
+    , dh_params_file_(dh_params_file)
+{
+    this->check_valid_<TlsKind::both>();
 }
 
 bool TlsConfiguration::is_active() const noexcept
 {
-    return false;
+    return this->kind_ != TlsKind::inactive;
+}
+
+template <TlsKind Kind>
+bool TlsConfiguration::compatible() const noexcept
+{
+    return this->kind_ != TlsKind::inactive &&
+           (Kind == this->kind_ or (this->kind_ == TlsKind::both && Kind != TlsKind::inactive));
```

Comment:
Be consistent using `or` or `||`.
*Note: I think `or` is not allowed in windows*

Reply:
right, win complains about it and I forgot to remove all of them